### PR TITLE
Protocol 3.3 and encrypted discovery

### PR DESCRIPTION
**Pull request by kueblc**
Preliminary support for protocol 3.3, including support for the new UDP broadcast mechanism.

Written to prioritize minimal change (hacky) and no breaking API changes. To be refactored in v6.0.0.

Tests pass. Additional testing was done with simulated 3.1 and 3.3 devices.

It wouldn't hurt to get a few more in vivo tests before merge. @Apollon77 @tsightler @kezakez @EricPalmquist @sylwester- @ChristianHoefer @thirug010

### CODE REVIEW

**Comment by Apollon77** [comment]
Coool. I‘m currently on vacation in tunesia but can give it a try in 1-2 weeks. So maybe all the others are faster then me ... ;-(

**Comment by Apollon77** [comment]
Code Review wise looks ok to me, but doc is missing for the new methods ;-)

I currently try to get such a device with new protocol so that I can test by myself too.

**Comment by Apollon77** [comment]
One more: was not the port for the tcp connection for the new device firmware also different? Or am I wrong with this?

**Comment by kueblc** [comment]
> Code Review wise looks ok to me, but doc is missing for the new methods ;-)

New methods should be considered internal for the time being, they will be refactored out in v6.0.0. Perhaps I can clarify this with an underscore?

> One more: was not the port for the tcp connection for the new device firmware also different? Or am I wrong with this?

It is the same, at least with the samples I have received. The UDP discovery process happens on a new port however.

**Comment by Apollon77** [comment]
Underscore is a Good convention for „private“ ... but in fact to start only with them maybe is confusing. Other idea is an @private as doc for the method. But in fact the same. 
So all ok ;-) (my point of view)

Thanks for the tcp port info.

**Comment by kueblc** [comment]
Actually turned out to be an easier refactor than I was anticipating, so I've taken out the protocol specific functions. Now the change set is isolated to payload treatment.

**Comment by EricPalmquist** [comment]
I tested with my device today - an outdoor 3 socket power strip and can read and write to all three.  You might consider in your constructor to show that the version must be specified as a string - it shows as an integer, which does not work.  Excellent work though - I appreciate your hard work.

**Comment by Apollon77** [comment]
Oh that it right. So we should adjust the check (use == ?) or add type conversion.

**Comment by kueblc** [comment]
> You might consider in your constructor to show that the version must be specified as a string - it shows as an integer, which does not work.

Good catch, though if you're using `.find()` you don't need to specify the version.

Should be fixed now. `MessageParser` now ensures `version` is a string on construction.

> Oh that it right. So we should adjust the check (use == ?) or add type conversion.

Funny I had used the `==` operator, but the linter complained, so I took it out to get tests passing without realizing the consequence. This linter drives me crazy sometimes, it doesn't like my coding style and fights with me on legitimate language features :stuck_out_tongue:

**Comment by thirug010** [comment]
Hi,
I have tested with 2 devices ( firmware 3.1.1 , 1.0.7) for both it is not finding the device with device id, with IP-Address it is able find it but it fails on 'data' with below error and it is not decrypted, Please let me know if i am missing something 

image

**Comment by kueblc** [comment]
> Hi,
> I have tested with 2 devices ( firmware 3.1.1 , 1.0.7) for both it is not finding the device with device id, with IP-Address it is able find it but it fails on 'data' with below error and it is not decrypted, Please let me know if i am missing something

Hi @thirug010, could you please post the test code you are using? If you are able to capture your network traffic please post this as well.

File: lib/cipher.js [hunk 1]
```diff
@@ -1,5 +1,7 @@
 const crypto = require('crypto');
 
+const UDP_KEY = crypto.createHash('md5').update('yGAdlopoPVldABfn', 'utf8').digest();
```
Comment (codetheweb) [inline lib/cipher.js]: Let's abstract the key out of the logic; maybe put it in a file called `config.json` or something?

File: index.js [hunk 2]
```diff
@@ -572,18 +599,14 @@ class TuyaDevice extends EventEmitter {
   async toggle(property = '1') {
     property = property.toString();
 
-    try {
-      // Get status
-      const status = await this.get({dps: property});
+    // Get status
+    const status = await this.get({dps: property});
```
Comment (codetheweb) [inline index.js]: Good catch, don't know why I put in a try/catch block there. 😛

File: index.js [hunk 3]
```diff
@@ -492,66 +497,88 @@ class TuyaDevice extends EventEmitter {
...
+        // Change protocol version if necessary
+        if (this.device.version !== dataRes.payload.version) {
           this.device.version = dataRes.payload.version;
 
-          // Cleanup
-          listener.close();
-          listener.removeAllListeners();
-          resolve(true);
+          // Update the parser
+          this.device.parser = new MessageParser({
```
Comment (codetheweb) [inline index.js]: So if a user passes in the IP manually to the constructor and don't use `.find()`, they'll have to pass in `version: 3.3` as well, correct?

If so, we'll need to add something to the docs about that.

**Comment by codetheweb** [comment]
Looks pretty good @kueblc; ready to merge once those changes are made.

I can't test it myself as I don't have any v3.3 devices, but I trust that everything works.

And if you give me some examples of bad linter rules, I can change the config. 😉

File: lib/cipher.js (hunk 1 above)
Comment (kueblc) [reply lib/cipher.js]: Sure no problem

File: index.js (hunk 2 above)
Comment (kueblc) [reply index.js]: :+1:

File: index.js (hunk 3 above)
Comment (kueblc) [reply index.js]: Yep, correct.

Comment (kueblc) [reply index.js]: Added a note to the README notes section

**Comment by kueblc** [comment]
> Looks pretty good @kueblc; ready to merge once those changes are made.

Ready when you are :+1: 

> And if you give me some examples of bad linter rules, I can change the config. wink

Cool, let's revisit for v6.0.0. Got a bunch of notes to go over.

**Comment by Apollon77** [comment]
I also currently try to get sich devices. Seems I need to purchase some :-(

It is not really about „bad“ Linter rules. In fact they are right and eg === is best practice also nimmt eyes but sometimes you want to have the „feature“ that == brings ;-) but in those rare cases you need to tell the Linter that you know what you did and to ignore it.

**Comment by thirug010** [comment]
@kueblc,

Please find the network logs ([logs.zip](https://github.com/codetheweb/tuyapi/files/3294685/logs.zip)) and I use the basic code [tuya-dev.txt](https://github.com/codetheweb/tuyapi/files/3294686/tuya-dev.txt), Please let me know if you need any other information

Thanks 
Thiru

**Comment by kueblc** [comment]
Thanks @thirug010, can you capture the network activity of the device? It looks like these are of your phone. You can use tcpdump or wireshark to create a pcap file.

Since `find` isn't working for you, I'd be interested to see what, if any, data is being broadcasted on UDP ports 6666-6668.

It also looks like you're passing `30` to `find`, which isn't valid

**Comment by tsightler** [comment]
I tested the same code as @thirug010 (tuya-dev.txt) with my switches and the 5.1.0 tuya-api and they worked fine with both find and IP.

Unfortunately my primary use case for tuyapi is to use it with [tuya-mqtt](https://github.com/TheAgentK/tuya-mqtt) which currently doesn't work with tuya-api 5.x although it looks like it should be easy to modify for that part.  However, looks like I'll also need to modify it for supporting 3.3 firmware since that code currently requires specifying the IP for each device (no support for find) but has no method to specify firmware version.

**Comment by AS137430** [comment]
@tsightler : https://github.com/TheAgentK/tuya-mqtt/blob/master/tuya-mqtt.js#L181 and https://github.com/codetheweb/tuyapi/issues/175#issuecomment-502475953.
Probably https://github.com/TheAgentK/tuya-mqtt/blob/master/tuya-mqtt.js#L125 needs to change too.

I have an (updated firmware) device new out from the box to test as well and have been waiting for @kueblc's changes. Thanks @kueblc and @codetheweb for your great work.

**Comment by tsightler** [comment]
Yeah, I actually hacked up tuya-mqtt last night to not require the IP and just use the find() capability and it worked just fine with the newer firmware.  Thanks to everyone for all the work here, I'm happy to be able to return to local control for my devices.

**Comment by AS137430** [comment]
@tsightler : Can you share your code please?

**Comment by kueblc** [comment]
Thanks @tsightler that's great to hear it worked out well for you. Your data was instrumental in understanding the new protocol.

**Comment by tsightler** [comment]
@AS137430 I will share the patched tuya-mqtt code in [my own tree](https://github.com/tsightler/tuya-mqtt) after flushing out a few remaining issues.  Mainly, while the code works to toggle the lights, it doesn't actually update the MQTT status as it seems like it's not correctly parsing the returned data.  I haven't had time to look at why but I suspect it's something pretty simple.

**Comment by thirug010** [comment]
> Thanks @thirug010, can you capture the network activity of the device? It looks like these are of your phone. You can use tcpdump or wireshark to create a pcap file.
> 
> Since `find` isn't working for you, I'd be interested to see what, if any, data is being broadcasted on UDP ports 6666-6668.
> 
> It also looks like you're passing `30` to `find`, which isn't valid

@kueblc,
I have checked with new devices + different Tuya based app now it is working without any issues. 
I can confirm it works for firmware version 1.0.7, 3.1.1 and 3.3 without using the firmware versions in constructors ( works for both Tuya switch, dimmer and fan light dimmer) 

Thanks 
Thiru

**Comment by tsightler** [comment]
@AS137430 I've published an initial version of tuya-mqtt that support 5.1.1 and discovery for 3.3 devices in my fork:  https://github.com/tsightler/tuya-mqtt
It seems to work for me but I only have simple, single switches from three different vendors and no lightbulbs so it's only tested in a small subset of possible use cases.  That being said, the overall changes are quite minimal.

**Pull request merged**